### PR TITLE
hkhanuja3 gaussian splatting

### PR DESCRIPTION
**Pull request by hkhanuja**
# Initial Implementation of Gaussian Splatting

## Initial Steps
Before being able to run the Gaussian Splatting, you need to perform the following steps (details for all these steps are in the main README):
* Clone the GTSfM repository
* Create a conda environment with the required dependencies
* Install GTSfM as a module
* Download the model weights
* Get the SfM points
* Convert the camera poses estimated by GTSfM to [nerfstudio](https://docs.nerf.studio/en/latest/) format:

## Additional Dependencies

* Check your torch version with the CUDA version on your system and set the appropriate paths.
* Install the additional necessary dependencies
```
pip install git+https://github.com/nerfstudio-project/gsplat.git
pip install git+https://github.com/rahul-goel/fused-ssim/ --no-build-isolation
pip install torchmetrics==1.7.3
pip install tyro==0.9.24
```

## Training

You can choose to start from random points by skipping the `init_type` argument. Other arguments can be altered similarly. By default, every sixth image will be part of the validation set; you can change it by adding `--test_every 10000` to your command. (basically an integer more than the number of images) 
```
CUDA_VISIBLE_DEVICES=0 python scripts/gaussian_splatting/custom_trainer.py default --data_dir results/nerfstudio_input --init_type sfm --max_steps 5000
```

## Rendering

In order to test the rendering function, a test `custom_traj_path` file is provided. You can render right after training by changing the training command by adding the ` --custom_traj_path scripts/gaussian_splatting/my_new_custom_path.json` or you can run the following command.

```
 CUDA_VISIBLE_DEVICES=0 python scripts/gaussian_splatting/custom_trainer.py default --data_dir results/nerfstudio_input --custom_traj_path scripts/gaussian_splatting/my_new_custom_path.json --ckpt results/custom_dataset/ckpts/ckpt_10000.pt
```

### CODE REVIEW

**Comment by dellaert** [comment]
Add some of PR comment in a READMe file?

File: scripts/gaussian_splatting/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Initial Implementation of Gaussian Splatting
+
+## Initial Steps
+Before being able to run the Gaussian Splatting, you need to perform the following steps (details for all these steps are in the main README):
+* Clone the GTSfM repository
+* Create a conda environment with the required dependencies
+* Install GTSfM as a module
+* Download the model weights
+* Get the SfM points
+* Convert the camera poses estimated by GTSfM to [nerfstudio](https://docs.nerf.studio/en/latest/) format:
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/README.md]: seems incomplete?

File: scripts/gaussian_splatting/README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+## Initial Steps
+Before being able to run the Gaussian Splatting, you need to perform the following steps (details for all these steps are in the main README):
+* Clone the GTSfM repository
+* Create a conda environment with the required dependencies
+* Install GTSfM as a module
+* Download the model weights
+* Get the SfM points
+* Convert the camera poses estimated by GTSfM to [nerfstudio](https://docs.nerf.studio/en/latest/) format:
+
+## Additional Dependencies
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/README.md]: this is okay for now, but ideally we want a requirements file or just add them to the dependencies in environment.yaml. they dont seem to be much.

File: scripts/gaussian_splatting/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+```
+pip install git+https://github.com/nerfstudio-project/gsplat.git
+pip install git+https://github.com/rahul-goel/fused-ssim/ --no-build-isolation
+pip install torchmetrics==1.7.3
+pip install tyro==0.9.24
+```
+
+## Training
+
+You can choose to start from random points by skipping the `init_type` argument. Other arguments can be altered similarly. By default, every sixth image will be part of the validation set; you can change it by adding `--test_every 10000` to your command. (basically an integer more than the number of images)
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/README.md]: 10000 is a bad example. Just mention the flags and what they mean.

File: scripts/gaussian_splatting/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+You can choose to start from random points by skipping the `init_type` argument. Other arguments can be altered similarly. By default, every sixth image will be part of the validation set; you can change it by adding `--test_every 10000` to your command. (basically an integer more than the number of images) 
+```
+CUDA_VISIBLE_DEVICES=0 python scripts/gaussian_splatting/custom_trainer.py default --data_dir results/nerfstudio_input --init_type sfm --max_steps 5000
+```
+
+## Rendering
+
+In order to test the rendering function, a test `custom_traj_path` file is provided. You can render right after training by changing the training command by adding the ` --custom_traj_path scripts/gaussian_splatting/my_new_custom_path.json` or you can run the following command.
+
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/README.md]: we dont need to render to a path, as its hard to say what a good path is. we could render to the test views though.

File: scripts/gaussian_splatting/custom_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    else:
+        raise ValueError(f"Unknown orientation method: {method}")
+    
+    transform = torch.cat([R, R @ -translation[..., None]], dim=-1)
+    poses_new = transform.to(poses.device) @ poses
+
+    return poses_new, transform
+
+# follows _undistort_image function from https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/data/utils/dataloaders.py
+def _undistort_image(
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/custom_dataloader.py]: I am not sure if this is actually useful. We dont estimate these parameters, so could potentially be useful if some dataset comes with it. None of our current datasets do I think.

File: scripts/gaussian_splatting/custom_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+class DataParserConfig:
+    """Simplified config for our custom parser, mimicking Nerfstudio's behavior."""
+    data: Path = Path()
+    scale_factor: float = 1.0
+    orientation_method: Literal["up"] = "up"
+    center_method: Literal["poses", "none"] = "poses"
+    auto_scale_poses: bool = True
+
+@dataclass
+class SimpleCameras:
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/custom_dataloader.py]: Has distortion, so definitely not so simple :) 
Maybe needs a better name.

File: scripts/gaussian_splatting/custom_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        self.fy *= scaling_factor
+        self.cx *= scaling_factor
+        self.cy *= scaling_factor
+        self.height = (self.height * scaling_factor).int()
+        self.width = (self.width * scaling_factor).int()
+
+    def __len__(self):
+        return self.camera_to_worlds.shape[0]
+
+class JsonParser:
```
Comment (akshay-krishnan) [inline scripts/gaussian_splatting/custom_dataloader.py]: I dont think we need to use the json files. this avoids the need to convert our results to nerfstudio format. 

we have i/o utils that provide everything needed here. Please take a look at https://github.com/borglab/gtsfm/blob/master/gtsfm/utils/io.py

Once you load the cameras from our results, you can scale / orient / center them using the same logic used here.

**Review by akshay-krishnan — COMMENTED** [review]
High level feedback, can do a second round for low-level once we address these. 

- the json format seems unnecessary
- please add docstrings at file, class and method levels. You can refer to oter files in gtsfm to get the format. 
- add type hints (also take a look at other files?)
- there is some duplication going on. I see some functions in both the trainer and dataloader. Move all common geometry-related functions to one of the util files. Preferable also write unit tests for these functions.

File: gtsfm/runner/gtsfm_runner_base.py [hunk 8]
```diff
@@ -121,6 +123,7 @@ def construct_argparser(self) -> argparse.ArgumentParser:
             "--share_intrinsics", action="store_true", help="Shares the intrinsics between all the cameras."
         )
         parser.add_argument("--mvs_off", action="store_true", help="Turn off dense MVS reconstruction")
+        parser.add_argument("--gs_off", action="store_true", help="Turn off Gaussian Splatting")
```
Comment (akshay-krishnan) [inline gtsfm/runner/gtsfm_runner_base.py]: okay, we want it default set to false. I know the mvs_off exists, but Ive wanted to change that for a while too. It'd be great if you could change both to `run_mvs` and `run_gs`

File: gtsfm/runner/gtsfm_runner_base.py [hunk 9]
```diff
@@ -282,10 +288,14 @@ def setup_ssh_cluster_with_retries(self) -> SSHCluster:
             )
         return cluster
 
-    def run(self) -> GtsfmData:
+    def run(self, graph_partitioner: GraphPartitionerBase = None) -> GtsfmData:
```
Comment (akshay-krishnan) [inline gtsfm/runner/gtsfm_runner_base.py]: self.graph_partitioner should exist. why this change?

File: gtsfm/runner/gtsfm_runner_base.py [hunk 10]
```diff
@@ -400,39 +411,29 @@ def run(self) -> GtsfmData:
...
         for idx, subgraph_result_dict in enumerate(subgraph_two_view_results):
-            logger.info(
-                f"Creating computation graph for subgraph {idx + 1}/{len(subgraph_two_view_results)} "
-                f"with {    len(subgraph_result_dict)} image pairs"
-            )
-            if len(subgraph_two_view_results) == 1:
-                # single partition
-                self.scene_optimizer.create_output_directories(None)
-            else:
-                self.scene_optimizer.create_output_directories(idx + 1)
```
Comment (akshay-krishnan) [inline gtsfm/runner/gtsfm_runner_base.py]: these should not be removed.

File: gtsfm/runner/gtsfm_runner_base.py [hunk 11]
```diff
@@ -469,23 +470,23 @@ def run(self) -> GtsfmData:
...
 
 
-def unzip_two_view_results(two_view_results: dict[tuple[int, int], TWO_VIEW_OUTPUT]) -> tuple[
-    dict[tuple[int, int], Rot3],
-    dict[tuple[int, int], Unit3],
-    dict[tuple[int, int], np.ndarray],
-    dict[tuple[int, int], TwoViewEstimationReport],
-    dict[tuple[int, int], TwoViewEstimationReport],
+def unzip_two_view_results(two_view_results: Dict[Tuple[int, int], TWO_VIEW_OUTPUT]) -> Tuple[
+    Dict[Tuple[int, int], Rot3],
```
Comment (akshay-krishnan) [inline gtsfm/runner/gtsfm_runner_base.py]: please revert back to using `dict` instead of `typing.Dict` . we dont need the typing dependency anymore since python 3.12. same for `tuple` and other data types

File: gtsfm/scene_optimizer.py [hunk 12]
```diff
@@ -25,14 +25,15 @@
 import gtsfm.utils.io as io_utils
 import gtsfm.utils.logger as logger_utils
 import gtsfm.utils.viz as viz_utils
+import gtsfm.splat.utils as gtsfm_utils
```
Comment (akshay-krishnan) [inline gtsfm/scene_optimizer.py]: `gtsfm_utils` is too generic, maybe splat_utils.

File: gtsfm/scene_optimizer.py [hunk 13]
```diff
@@ -71,29 +72,29 @@ def __init__(
...
         multiview_optimizer: MultiViewOptimizer,
         dense_multiview_optimizer: Optional[MVSBase] = None,
+        gaussian_splatting_optimizer: Optional[GSBase] = None,
         save_two_view_correspondences_viz: bool = False,
         save_3d_viz: bool = False,
         save_gtsfm_data: bool = True,
         pose_angular_error_thresh: float = 3,  # in degrees
         output_root: str = DEFAULT_OUTPUT_ROOT,
         output_worker: Optional[str] = None,
-        graph_partitioner: Optional[GraphPartitionerBase] = SinglePartition(),
```
Comment (akshay-krishnan) [inline gtsfm/scene_optimizer.py]: shouldnt be removed.

File: gtsfm/scene_optimizer.py [hunk 14]
```diff
@@ -71,29 +72,29 @@ def __init__(
...
         self._save_two_view_correspondences_viz = save_two_view_correspondences_viz
         self._save_3d_viz = save_3d_viz
-        self.run_dense_optimizer = self.dense_multiview_optimizer is not None
 
         self._save_gtsfm_data = save_gtsfm_data
         self._pose_angular_error_thresh = pose_angular_error_thresh
         self.output_root = Path(output_root)
         self._output_worker = output_worker
-        self.graph_partitioner = graph_partitioner
+        self._create_output_directories()
```
Comment (akshay-krishnan) [inline gtsfm/scene_optimizer.py]: unnecessary change?

File: gtsfm/scene_optimizer.py [hunk 15]
```diff
@@ -103,30 +104,25 @@ def __repr__(self) -> str:
...
-        """Create plot base path."""
-        plot_base_path = self.output_root / "plots"
-        os.makedirs(plot_base_path, exist_ok=True)
-        return plot_base_path
-
-    def create_output_directories(self, partition_index: Optional[int]) -> None:
+    def _create_output_directories(self) -> None:
         """Create various output directories for GTSFM results, metrics, and plots."""
-        # Construct subfolder if partitioned
-        partition_folder = f"partition_{partition_index}" if partition_index is not None else ""
```
Comment (akshay-krishnan) [inline gtsfm/scene_optimizer.py]: we need this partition folder

File: gtsfm/scene_optimizer.py [hunk 16]
```diff
@@ -103,30 +104,25 @@ def __repr__(self) -> str:
...
+        self._metrics_path = self.output_root / "result_metrics"
+        self._results_path = self.output_root / "results"
 
         # plot paths
         self._plot_correspondence_path = self._plot_base_path / "correspondences"
         self._plot_ba_input_path = self._plot_base_path / "ba_input"
         self._plot_results_path = self._plot_base_path / "results"
         self._mvs_ply_save_fpath = self._results_path / "mvs_output" / "dense_pointcloud.ply"
 
+        self._gaussian_splatting_path = self._results_path/ "gaussian_splatting_output"
```
Comment (akshay-krishnan) [inline gtsfm/scene_optimizer.py]: for brevity, maybe just use `splat` or `gs` instead of `gaussian_splatting`

File: gtsfm/splat/gaussian_splatting.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
+import json
```
Comment (akshay-krishnan) [inline gtsfm/splat/gaussian_splatting.py]: we need file-level docstrings, PTAL at other files.

File: gtsfm/splat/gaussian_splatting.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+import itertools
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import tqdm
+import tyro
```
Comment (akshay-krishnan) [inline gtsfm/splat/gaussian_splatting.py]: we do we need tyro? do we already have that dependency?

File: gtsfm/splat/gaussian_splatting.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+            features_dc = torch.nn.Parameter(sh0_values)
+        else:
+            if init_type == "sfm":
+                print("Warning: 'sfm' init chosen but no points found. Falling back to 'random'.")
+            print(f"Initializing with {init_num_pts} random points.")
+            random_scale = 10.0
+            points = torch.nn.Parameter((torch.rand((init_num_pts, 3)) - 0.5) * random_scale)
+            features_dc = torch.nn.Parameter(torch.rand(init_num_pts, 3))
+
+        print("Calculating initial scales based on nearest neighbors...")
```
Comment (akshay-krishnan) [inline gtsfm/splat/gaussian_splatting.py]: use logger instead of print, also update elsewhere

File: gtsfm/splat/utils.py [hunk 20]
```diff
@@ -0,0 +1,240 @@
+import os
```
Comment (akshay-krishnan) [inline gtsfm/splat/utils.py]: this file should move to gtsfm/utils/splat.py

**Review by akshay-krishnan — COMMENTED** [review]
I see files / classes / methods without docstrings. we need them for all three. Also looks like youre overwriting changes from master, which looks unintentional. I havent reviewed the GS code in detail.
